Transition to activity service for viewer counts

spladug opened the pull request:
This transitions to the activity service and is intended for each commit to be deployed separately. 

Later, we can move off the liveupdate pixel endpoint to the activity service gateway to offload that aspect from the r2 stack.

:eyeglasses: @dellis23 @bsimpson63

spladug committed:
Stop writing activity to Cassandra

bsimpson63 commented:
> Because of the differences involved, we can only use the activity service for still-open live threads.

Why is this?

bsimpson63 commented:
> The Cassandra implementation has been drastically undercounting.

I forget, did you figure out why this was happening?

spladug commented:
>> Because of the differences involved, we can only use the activity service for still-open live threads.

> Why is this?

The Cassandra implementation used TTL'd columns and would just iterate over all rows in the activity CF. This was a cheap and easy way of knowing "who is currently active at all". Now that we don't have that, we'd need to either: 1) iterate over all existing live threads and check their activity, or 2) store in redis a "currently active" set somehow. I opted for (1) with the optimization of focusing only on threads in the live state. 

>> The Cassandra implementation has been drastically undercounting.

> I forget, did you figure out why this was happening?

No :(.  I tried increasing the WCL to no effect. Not really sure beyond that what might be going on. The subreddit implementation had pretty good accuracy, so I don't think it was the C* data model itself, but the haproxy logs of the pixel hits here were drastically different from the numbers being reported on the site.

bsimpson63 commented:
how many rows are in the cf? can we do a full range query forever?
On Apr 27, 2016 10:05 AM, "Neil Williams" <notifications@github.com> wrote:

> Because of the differences involved, we can only use the activity service
> for still-open live threads.
>
> Why is this?
>
> The Cassandra implementation used TTL'd columns and would just iterate
> over all rows in the activity CF. This was a cheap and easy way of knowing
> "who is currently active at all". Now that we don't have that, we'd need to
> either: 1) iterate over all existing live threads and check their activity,
> or 2) store in redis a "currently active" set somehow. I opted for (1) with
> the optimization of focusing only on threads in the live state.
>
> The Cassandra implementation has been drastically undercounting.
>
> I forget, did you figure out why this was happening?
>
> No :(. I tried increasing the WCL to no effect. Not really sure beyond
> that what might be going on. The subreddit implementation had pretty good
> accuracy, so I don't think it was the C* data model itself, but the haproxy
> logs of the pixel hits here were drastically different from the numbers
> being reported on the site.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/reddit/reddit-plugin-liveupdate/pull/130#issuecomment-215149902>
>

spladug commented:
> how many rows are in the cf? can we do a full range query forever?

Currently 25K and it takes 5.5 seconds to iterate through all of them. We have some room to grow there and if it becomes a pain we can just split up the cron job to look at specific ranges.

bsimpson63 commented:
:fish:
On Apr 27, 2016 8:22 AM, "Neil Williams" <notifications@github.com> wrote:

> This transitions to the activity service and is intended for each commit
> to be deployed separately.
>
> Later, we can move off the liveupdate pixel endpoint to the activity
> service gateway to offload that aspect from the r2 stack.
>
> [image: :eyeglasses:] @dellis23 <https://github.com/dellis23> @bsimpson63
> <https://github.com/bsimpson63>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/reddit/reddit-plugin-liveupdate/pull/130
> Commit Summary
>
>    - Write visits to the activity service
>    - Read activity counts from the activity service
>    - Stop writing activity to Cassandra
>
> File Changes
>
>    - *M* reddit_liveupdate/activity.py
>    <https://github.com/reddit/reddit-plugin-liveupdate/pull/130/files#diff-0>
>    (72)
>    - *M* reddit_liveupdate/controllers.py
>    <https://github.com/reddit/reddit-plugin-liveupdate/pull/130/files#diff-1>
>    (9)
>    - *M* reddit_liveupdate/models.py
>    <https://github.com/reddit/reddit-plugin-liveupdate/pull/130/files#diff-2>
>    (21)
>    - *M* reddit_liveupdate/pages.py
>    <https://github.com/reddit/reddit-plugin-liveupdate/pull/130/files#diff-3>
>    (10)
>    - *M* reddit_liveupdate/public/static/js/liveupdate/statusBar.js
>    <https://github.com/reddit/reddit-plugin-liveupdate/pull/130/files#diff-4>
>    (7)
>    - *M* reddit_liveupdate/templates/liveupdateevent.html
>    <https://github.com/reddit/reddit-plugin-liveupdate/pull/130/files#diff-5>
>    (4)
>
> Patch Links:
>
>    - https://github.com/reddit/reddit-plugin-liveupdate/pull/130.patch
>    - https://github.com/reddit/reddit-plugin-liveupdate/pull/130.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/reddit/reddit-plugin-liveupdate/pull/130>
>

kjoconnor commented:
I'm just slightly worried about having this go out and continue writing to our ElastiCache set up since we already have some pretty bad issues with evictions.

dellis23 commented:
@kjoconnor Can you explain the relationship to ElastiCache? (I know nothing about it)

kjoconnor commented:
We use ElastiCache for the Redis servers that back the activity service.

dellis23 commented on reddit_liveupdate/activity.py:
```diff
@@ -18,51 +19,48 @@
+
+        try:
+            with c.activity_service.retrying() as svc:
+                infos = svc.count_activity_multi(context_ids.keys())
```
I'm having difficulty figuring out what's going on here (and can't find `count_activity_multi` in the live update plugin or the activity service).  It seems like we're kicking off a count process for every live update thread?  Perhaps a comment would help.

spladug replied:
You won't find it in the activity service yet, it's in this pull request https://github.com/reddit/reddit-service-activity/pull/7

We're just fetching the count of activity for each event, but doing it in a big batch.

dellis23 replied:
Ah, I see.  Thanks.

dellis23 commented on reddit_liveupdate/activity.py:
```diff
@@ -18,51 +19,48 @@
-        if is_active:
             try:
-                count = ActiveVisitorsByLiveUpdateEvent.get_count(event_id)
+                LiveUpdateActivityHistoryByEvent.record_activity(
```
Do we keep track of this indefinitely?  How often are we adding these?  (Not saying it's bad or should be documented; just for own curiosity).

spladug replied:
Yes indefinitely. Once per cron job which means once per minute.

dellis23 commented on reddit_liveupdate/activity.py:
```diff
@@ -18,51 +19,48 @@
-            event_counts[event_id] = count
+            try:
+                event = LiveUpdateEvent.update_activity(
+                    event_id, info.count, info.is_fuzzed)
```
What determines whether the count is fuzzed?  I can't find the method that returns the data to dig in.

spladug replied:
https://github.com/reddit/reddit-service-activity/blob/master/reddit_service_activity/__init__.py#L23-L30

dellis23 commented:
🐟  A few questions, but looks good!

spladug commented:
:haircut: only the last commit is new. the others moved down 'cus i had to rebase. sorry :(

i screwed up in testing and didn't verify the activity was actually recorded. the service was not present because it's not using the full reddit controller.

dellis23 commented on reddit_liveupdate/controllers.py:
```diff
@@ -195,7 +195,11 @@ def _pixel_contents(self):
         return self._pixel_data
 
-    def GET_pixel(self, event):
+    # this decorator takes **kwargs which routes treats specially and throws
```
🔕 Haha, it took me a second to parse the words "routes treats specially", not realizing "routes" was a noun rather than a verb.

spladug replied:
i should go german and capitalize it :P

dellis23 commented:
Re-:fish:

bsimpson63 commented:
🐟

spladug committed:
activity: Filter out closed events before chunking

This ensures that we get full-sized chunks since so many events are
long-since completed. Currently, chunks of a hundred events are being
filtered down to just three or four before we actually do any batch
lookups. This seriously hampers the effectiveness of the batching.

spladug commented:
:haircut: only the last commit is new now. at this point all has been deployed but the "stop writing to cassandra" commit and the new one.

bsimpson63 commented:
🐟

dellis23 commented:
🐟  So is the chunking less for the query and more for the processing that comes after?  I'm assuming so since you're exhausting the iterator now.

spladug commented:
yes, the chunking is just for the `count_activity_multi`

spladug commented:
b8c2193db5a0f74c09c9ed13ea3b92a96a1c1bf2...841772d6efe939bf58db4a01bb3e4776475ec8e4
